src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+				}
+				return -EPROTO;
+			case RATBAG_MACRO_EVENT_WAIT:
+				return -EPROTO;
+			default:
+				return -EINVAL;
+		}
+	}
+
+	return -EINVAL;
+}
+
+/*
+ * Converts raw macro events to ratbag macro and writes to a button
+ *
+ * @param raw A buffer of macro events at least HOLTEK8_MAX_MACRO_EVENTS of size
+ * @return 0 on success or a negative errno.
+ */
+static int
+holtek8_macro_from_events(struct ratbag_button *button, const struct holtek8_macro_event *raw)
+{
+	struct ratbag_device *device = button->profile->device;
+	struct holtek8_data *drv_data = device->drv_data;
+	struct ratbag_button_macro *macro = ratbag_button_macro_new("macro");
+	int i, rc;
+	int macro_i = 0;
+	unsigned int delay = 0;
+	unsigned int delay_base_ms;
+	unsigned int key;
+
+	switch (drv_data->api_version) {
+		case HOLTEK8_API_A:
+			delay_base_ms = HOLTEK8A_MACRO_DELAY_MS;
+			break;
+		case HOLTEK8_API_B:
+			delay_base_ms = HOLTEK8B_MACRO_DELAY_MS;
+			break;
+	}
+
+	button->action.type = RATBAG_BUTTON_ACTION_TYPE_UNKNOWN;
+
+	for (i = 0; i < HOLTEK8_MAX_MACRO_EVENTS; i++) {
+		const struct holtek8_macro_event *event = &raw[i];
+
+		if (event->delay == 0 && event->release == 0 && event->key == 0)
+			break;
+
+		if (macro_i >= MAX_MACRO_EVENTS)
+			goto overflow;
+
+		if (event->key == HOLTEK8_MACRO_CMD_WAIT) {
+			if (event->delay != 0 || event->release != 0) {
+				rc = -EINVAL;
+				goto err;
+			}
+			if (++i >= HOLTEK8_MAX_MACRO_EVENTS) {
+				rc = -ENODATA;
+				goto err;
+			}
+
+			delay += get_unaligned_be_u16((uint8_t*) &raw[i]) * 2;
+			continue;
+		}
+		if (event->key == HOLTEK8_MACRO_CMD_MOUSE) {
+			//no support in ratbag for mouse movements in macros
+			i += 1;
+			continue;
+		}
+
+		if (delay > 1) {
+			ratbag_button_macro_set_event(macro, macro_i++, RATBAG_MACRO_EVENT_WAIT, delay * delay_base_ms);
+			if (macro_i >= MAX_MACRO_EVENTS)
+				goto overflow;
+		}
+		delay = event->delay;
+
+		switch (event->key) {
+			case HOLTEK8_BUTTON_MOUSE_LEFT:   key = BTN_LEFT; break;
+			case HOLTEK8_BUTTON_MOUSE_RIGHT:  key = BTN_RIGHT; break;
+			case HOLTEK8_BUTTON_MOUSE_MIDDLE: key = BTN_MIDDLE; break;
+			case HOLTEK8_BUTTON_MOUSE_MB4:    key = BTN_SIDE; break;
+			case HOLTEK8_BUTTON_MOUSE_MB5:    key = BTN_EXTRA; break;
+			default:
+				key = ratbag_hidraw_get_keycode_from_keyboard_usage(device, event->key);
+		}
+
+		ratbag_button_macro_set_event(macro, macro_i++, event->release ? RATBAG_MACRO_EVENT_KEY_RELEASED : RATBAG_MACRO_EVENT_KEY_PRESSED, key);
+	}
+
+	ratbag_button_copy_macro(button, macro);
+	ratbag_button_macro_unref(macro);
+	return 0;
+
+overflow:
+	log_error(device->ratbag, "Can't fit device macro for button %u in ratbag macro\n", button->index);
+	rc = -EOVERFLOW;
+	ratbag_button_copy_macro(button, macro);
+err:
+	ratbag_button_macro_unref(macro);
+	return rc;
+}
+
+/*
+ * Reads ratbag macro from a button and converts to raw macro events
+ *
+ * @param raw A buffer of macro events at least HOLTEK8_MAX_MACRO_EVENTS of size
```

Comment:
Same about buffer size:
[Jan 28, 2024](https://github.com/libratbag/libratbag/pull/1561/files#r1468903332)
I would prefer passing buffer size along with the pointer, I think it's a bit less error-prone.